docs/user-defined-type-functions.md

```diff
@@ -38,7 +38,7 @@ type Person = {
 type ty = rawget<Person, "name"> -- ty = string
 ```
 
-Type functions operate on two stages: type analysis and runtime. They are _called_ at the type level of Luau, but _runs_ at runtime to operate on the values of types (such as the arguments of the function). For this reason, type functions are declared using parenthesis and invoked using angle brackets. Moreover, every type functions will be part of and ran within the same scope (rlly? what if they are defined in a function) to allow type functions to call other type functions. We will not (and probably never) allow type functions to call regular functions for the sake of sandboxing runtime and analysis time.
+Type functions operate on two stages: type analysis and runtime. They are _called_ at the type level of Luau, but _runs_ at runtime to operate on the values of types (such as the arguments of the function). For this reason, type functions are declared using parenthesis and invoked using angle brackets. Moreover, every type functions will be part of and ran within the same scope to allow type functions to call other type functions. We will not (and probably never) allow type functions to call regular functions for the sake of sandboxing the runtime and analysis time. There will also be a user-configurable timeout for type functions based on the number of instructions. For Roblox, this could be done by editing the config file for RCC and Studio and for non-Roblox users, this could be done by setting a flag when running the compiler.
```

Comment:
How will this instruction limit be implemented?

Reply:
I there is an option to force single-stepping the interpreter using an option commonly used for debuggers.

It should be mentioned why the time limit is not used (it's used for time limits in other runtime scenarios).

While it could be said that time limit is too dependent on the CPU, instruction limit is also not fixed as updates to the compiler can change the bytecode and cause previously typechecked code to fail (should be noted).

Reply:
ah, the halting problem, our old friend.

Reply:
I think instruction count is ok. The issue with doing wall clock measurement is that it's dependent on CPU. At least with instruction count it's guaranteed to be consistent across CPUs, even though it's not guaranteed to be consistent across compiler versions. This is nevertheless an improvement over wall clock timeout.

Reply:
Yeah, having used some type systems that rely on SMT solvers with a wall clock timeout, I really don't want to deal with inconsistencies across CPUs, and an amorphous target for developers of "optimize this to be fast enough always." At least with instruction counts, if it breaks, there is a very clear line for what you need to do to fix it again.

Reply:
update: an instruction limit will not be implemented. type functions will not allow any constructs that can cause infinite runtime

Reply:
Unfortunately we're gonna have to go back to doing instruction limit. It's not as simple as "make sure certain AST nodes doesn't occur in this context."

```luau
for i = 1, math.huge do
end
```
```lua
local t = {1}
for _ in t do
  table.insert(t, 1)
end
```

Reply:
I think I want to see wordings that essentially says "_potentially_ infinite loops in type function context are not supported." The idea I want to get across is that we reserve the right to modify the runtime semantics at any point that makes type functions throw a runtime error if infinite loops are found, which would cause these type functions to not reduce.

Meaning that at some point, we may add code that detects whether `table.insert` is being used on a table with an active iterator, and have them throw a runtime error.